### PR TITLE
Add the sriov network operator ci project

### PR DESCRIPTION
**Pull request by abdallahyas**
Add the sriov_network_operator project to test the e2e tests of
the sriov-network-operator on nvidia hardware.

### CODE REVIEW

**Comment by abdallahyas** [comment]
Patch is tested and should be working, but it needs wrapping up (docs and nits here and there)

**Comment by Mmduh-483** [comment]
Can you split the commits as follow:
- Updating prepare_setup.sh script
- Install and deploy kind
- Background script to map a kind node with PF and its VFs
- Add sriov network operator CI
- Update README

**Review by e0ne — APPROVED** [review]

File: prepare_setup.sh
Comment (Mmduh-483) [inline prepare_setup.sh]: Use github api to get the latest release `https://api.github.com/repos/kubernetes-sigs/kind/releases/latest` , it returns a json, `result['name']`

File: prepare_setup.sh
Comment (Mmduh-483) [inline prepare_setup.sh]: use `wget -O` instead of `mv`

File: common/common_functions.sh
Comment (Mmduh-483) [inline common/common_functions.sh:889]: Why is this needed?

File: common/common_functions.sh
Comment (Mmduh-483) [inline common/common_functions.sh]: You didn't specify the cluster name

File: deploy/vf-switcher.service
Comment (Mmduh-483) [inline deploy/vf-switcher.service:1]: Why need the service, vf switcher is a background job it is not needed at boot

File: sriov_network_operator/sriov_network_operator_ci_start.sh
Comment (Mmduh-483) [inline sriov_network_operator/sriov_network_operator_ci_start.sh:84]: you didn't specify the operator `SRIOV_NETWORK_OPERATOR_IMAGE ` and daemon `SRIOV_NETWORK_CONFIG_DAEMON_IMAGE ` images and the cluster type `CLUSTER_TYPE`

sriov operator uses latest master if they are not speicfied and cluster type to openshift

File: sriov_network_operator/README.md
Comment (Mmduh-483) [inline sriov_network_operator/README.md]: Please add more details about the scenario tested and the passed parameters

File: prepare_setup.sh
Comment (abdallahyas) [reply prepare_setup.sh]: IMO, pinning it to a certain version would be best to not be surprised of different behavior when preparing new nodes, though we will need to figure out a way to install latest stable Kubenetes in the node image.

Comment (Mmduh-483) [reply prepare_setup.sh]: > IMO, pinning it to a certain version would be best to not be surprised of different behavior when preparing new nodes, though we will need to figure out a way to install latest stable Kubenetes in the node image.

In k8s ci we are using the latest official release to find if there will be any surprises later and I belive it should be the same for kind

File: common/common_functions.sh
Comment (abdallahyas) [reply common/common_functions.sh:889]: This is need to be able to create VFs, since to create them, write access to the sriov_numvfs is needed. and kind by default remount the sys fs to be read only. i think they do so to not have many container (nodes) writing to the same file.

File: deploy/vf-switcher.service
Comment (abdallahyas) [reply deploy/vf-switcher.service:1]: using services, it would be easier to start, stop, and collect logs across scripts. It would also be easier to debug in case of failures.

File: sriov_network_operator/sriov_network_operator_ci_start.sh
Comment (abdallahyas) [reply sriov_network_operator/sriov_network_operator_ci_start.sh:84]: see [Makefile](https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/master/Makefile#L195-L200) for basic env variables, and see [vars](https://github.com/Mellanox/kubernetes-ci/pull/105/files?file-filters%5B%5D=.sh#diff-1da590bb51c16903726785c833421c5daa35025cdc36c923aca64459aae22f1cR17-R21) for the image variables

File: prepare_setup.sh
Comment (abdallahyas) [reply prepare_setup.sh]: That is true for deployment components, but kind is a tool we use to deploy those components, so i think pinning kind and building the node image with latest stable Kubernetes is the correct way of doing it. since also latest kind does not necessarily deploy the latest Kubernetes.

File: sriov_network_operator/README.md
Comment (abdallahyas) [reply sriov_network_operator/README.md]: Done

File: common/common_functions.sh
Comment (abdallahyas) [reply common/common_functions.sh]: added it

File: prepare_setup.sh
Comment (abdallahyas) [reply prepare_setup.sh]: Done

File: common/common_functions.sh
Comment (Mmduh-483) [reply common/common_functions.sh:889]: Is it configurable in kind to mount it with write permission?

File: sriov_network_operator/sriov_network_operator_ci_start.sh
Comment (Mmduh-483) [reply sriov_network_operator/sriov_network_operator_ci_start.sh:84]: I see you set the env vars for images, what about the cluster type?

File: common/common_functions.sh
Comment (abdallahyas) [reply common/common_functions.sh:889]: no, it is hard coded, see https://github.com/kubernetes-sigs/kind/blob/b6bc112522651d98c81823df56b7afa511459a3b/images/base/files/usr/local/bin/entrypoint#L147

File: sriov_network_operator/sriov_network_operator_ci_start.sh
Comment (abdallahyas) [reply sriov_network_operator/sriov_network_operator_ci_start.sh:84]: it is set in the Makefile of the sriov-network operator

**Review by Mmduh-483 — APPROVED** [review]

File: common/common_functions.sh
Comment (ykulazhenkov) [inline common/common_functions.sh]: nit: you mixing "[" and "[[" syntax for the same checks. It is better to use consistent style

File: common/common_functions.sh
Comment (ykulazhenkov) [inline common/common_functions.sh]: It looks like here is lack of `status` initialization `local status=0`

Should be `let status=status+$?` replaced with `let status=$status+$`?

File: common/common_functions.sh
Comment (ykulazhenkov) [inline common/common_functions.sh]: `let status=status+$?` >> `let status=$status+$`

File: common/common_functions.sh
Comment (ykulazhenkov) [reply common/common_functions.sh:889]: I think it is possible to use more specific paths in kind config, then it should work, eg
```
   extraMounts:
    - hostPath: /sys/class/net
      containerPath: /sys/class/net
    - hostPath: /sys/devices
      containerPath: /sys/devices
```

File: common/common_functions.sh
Comment (ykulazhenkov) [inline common/common_functions.sh:1104]: It is better to move addition of cluster name to the commit where you add upload_image_to_kind function

File: deploy/vf-netns-switcher.sh
Comment (ykulazhenkov) [inline deploy/vf-netns-switcher.sh]: confusing indentation, tab used instead of spaces

File: deploy/vf-netns-switcher.sh
Comment (ykulazhenkov) [inline deploy/vf-netns-switcher.sh]: "pid" > "$pid" ?

File: deploy/vf-netns-switcher.sh
Comment (ykulazhenkov) [inline deploy/vf-netns-switcher.sh]: subshell $() missing

File: common/common_functions.sh
Comment (ykulazhenkov) [inline common/common_functions.sh]: I think you should move this to somewhere else. Function with name "prepare_vf_switcher_confs" should not remove symlink for container's ns.

File: prepare_setup.sh
Comment (ykulazhenkov) [inline prepare_setup.sh]: Please, move "mv to -O" fix to the commit where you add kind_install function

File: deploy/vf-netns-switcher.sh
Comment (ykulazhenkov) [inline deploy/vf-netns-switcher.sh]: I can't see usage of `--vf-driver <> --vf-count <>`

File: deploy/vf-netns-switcher.sh
Comment (ykulazhenkov) [inline deploy/vf-netns-switcher.sh]: Maybe we can drop this? It look like we have no keys with "-" in the config at the moment.

File: deploy/vf-netns-switcher.sh
Comment (ykulazhenkov) [inline deploy/vf-netns-switcher.sh]: Maybe it will be better to use something like this?
```basename $(readlink /sys/class/net/enp5s0f0/device)```

File: deploy/vf-netns-switcher.sh
Comment (ykulazhenkov) [inline deploy/vf-netns-switcher.sh]: unused function

File: deploy/vf-netns-switcher.sh
Comment (ykulazhenkov) [inline deploy/vf-netns-switcher.sh]: Maybe it will be better to use here `ip netns e <ns> <cmd>` instead of docker exec?

File: sriov_network_operator/sriov_network_operator_ci_test.sh
Comment (ykulazhenkov) [inline sriov_network_operator/sriov_network_operator_ci_test.sh]: All these variables are needed here?

File: common/common_functions.sh
Comment (abdallahyas) [reply common/common_functions.sh:889]: @ykulazhenkov  unfortunately this will not work also, since this basically mount the directory inside the container, then the entrypoint will do the remount to read only after the container is started.

P.S. I did try this the first time i faced the issue, but that did not work.

File: deploy/vf-netns-switcher.sh
Comment (ykulazhenkov) [inline deploy/vf-netns-switcher.sh]: nit: in this line tab used instead of spaces for indentation

File: deploy/vf-netns-switcher.sh
Comment (ykulazhenkov) [inline deploy/vf-netns-switcher.sh]: nit: tab + spaces used for this line indentation

File: common/common_functions.sh
Comment (ykulazhenkov) [inline common/common_functions.sh:849]: It looks like this call can fail after this PR  [1] will be merged 
I think it is better to discuss with @Mmduh-483 which PR should be merged first.
[1] https://github.com/Mellanox/kubernetes-ci/pull/110/files

**Review by ykulazhenkov — APPROVED** [review]

File: deploy/vf-netns-switcher.sh
Comment (abdallahyas) [reply deploy/vf-netns-switcher.sh]: thanks for the review and sorry for these small nits, i will investigate if we can have some kind of CI that would do customized linting for bash.

File: deploy/vf-netns-switcher.sh
Comment (abdallahyas) [reply deploy/vf-netns-switcher.sh]: removed

File: common/common_functions.sh
Comment (abdallahyas) [reply common/common_functions.sh:849]: you are correct, phases that should not run will run now. @Mmduh-483 WDYT?

Comment (Mmduh-483) [reply common/common_functions.sh:849]: I think we can add a flag `--phases` to run specific phases and calling this flag will ignore `--skip-phases` for simplicty

Comment (Mmduh-483) [reply common/common_functions.sh:849]: Check PR #112

Comment (abdallahyas) [reply common/common_functions.sh:849]: updated the PR, can you have a look

Comment (Mmduh-483) [reply common/common_functions.sh:849]: You need to run `prepare-ci-environment` phase also. `deploy-kind/create-cluster` sub-phase renders the config file then try to move it to the artifacts dir that created in `prepare-ci-environment`

Comment (abdallahyas) [reply common/common_functions.sh:849]: @Mmduh-483 I do that manually in the start script. see https://github.com/Mellanox/kubernetes-ci/pull/105/files#diff-1da590bb51c16903726785c833421c5daa35025cdc36c923aca64459aae22f1cR110

**Review by e0ne — APPROVED** [review]
LGTM

**Review by ykulazhenkov — APPROVED** [review]

**Pull request merged**
